fastparse/shared/src/main/scala/fastparse/Api.scala

```diff
@@ -1,51 +1,130 @@
+
+  val ElemPred = Intrinsics.ElemPred[ElemType, Repr] _
+  def ElemIn(seqs: Seq[ElemType]*) = Intrinsics.ElemIn[ElemType, Repr](seqs.map(_.toIndexedSeq): _*)
+  def ElemsWhile(pred: ElemType => Boolean, min: Int = 1) = Intrinsics.ElemsWhile[ElemType, Repr](pred, min)
+  def SeqIn(seqs: Seq[ElemType]*) = Intrinsics.StringIn[ElemType, Repr](seqs.map(_.toIndexedSeq): _*)
 
   val NoTrace = parsers.Combinators.NoTrace
   val NoCut = parsers.Combinators.NoCut
 
   val & = parsers.Combinators.Lookahead
 
-  implicit def wspStr(s: String): P0 =
-    if (s.length == 1) parsers.Terminals.CharLiteral(s(0))
-    else parsers.Terminals.Literal(s)
-
   def P[T](p: => Parser[T])(implicit name: sourcecode.Name): Parser[T] =
     parsers.Combinators.Rule(name.value, () => p)
 
   type P0 = Parser[Unit]
-  type Parser[+T] = core.Parser[T]
+  type Parser[+T] = core.Parser[T, ElemType, Repr]
   type P[+T] = Parser[T]
-  val ParseError = core.ParseError
-  type ParseError = core.ParseError
+  val ParseError = core.ParseError[ElemType] _
+  type ParseError = core.ParseError[ElemType]
+}
+
+trait StringApi extends Api[Char, String] {
+
+  val elemHelper = implicitly[ElemHelper[Char]]
+  val parserHelper = implicitly[ParserHelper[Char]]
+  val ordering = implicitly[Ordering[Char]]
+
+  val AnyChar = AnyElem
+
+  val CharPred = ElemPred
+  def CharIn(strings: Seq[Char]*) = ElemIn(strings: _*)
+  def CharsWhile(pred: Char => Boolean, min: Int = 1) = ElemsWhile(pred, min)
+  def StringIn(strings: Seq[Char]*) = SeqIn(strings: _*)
+
+  val CharPredicates = fastparse.CharPredicates
+  val IgnoreCase = parsers.Terminals.IgnoreCase[String] _
+
+  implicit def wspStr(s: String): P0 =
+    if (s.length == 1) parsers.Terminals.ElemLiteral(s(0))
+    else parsers.Terminals.Literal(s)
 }
-object all extends Api{
-  implicit def parserApi[T, V](p: T)(implicit c: T => core.Parser[V]): ParserApi[V] =
-    new ParserApiImpl[V](p)
+
+object allString extends StringApi{
+  implicit def parserApi[T, V](p: T)(implicit c: T => core.Parser[V, Char, String]): ParserApi[V, Char, String] =
+    new ParserApiImpl[V, Char, String](p)
+}
+object emptyStringApi extends StringApi
+
+
+trait ByteApi extends Api[Byte, Array[Byte]] {
+
+  val AnyByte = AnyElem
+
+  val elemHelper = implicitly[ElemHelper[Byte]]
+  val parserHelper = implicitly[ParserHelper[Byte]]
+  val ordering = implicitly[Ordering[Byte]]
+
+  val BytePred = ElemPred
+  def ByteIn(seqs: Seq[Byte]*) = ElemIn(seqs: _*)
+  def BytesWhile(pred: Byte => Boolean, min: Int = 1) = ElemsWhile(pred, min)
+
+  def ByteSeq[T](bytes: T*)(implicit num: Numeric[T]) = bytes.map(b => num.toInt(b).toByte).toArray
+  def BS[T](bytes: T*)(implicit num: Numeric[T]) = ByteSeq[T](bytes: _*)
+  type ByteSeq = Array[Byte]
+  type BS = ByteSeq
+
+  implicit def wspByteSeq(seq: Array[Byte]): P0 =
+    if (seq.length == 1) parsers.Terminals.ElemLiteral(seq(0))
+    else parsers.Terminals.Literal(seq)
+
+  implicit class ByteSeqUtils(s: String) {
```

Comment:
Can we implement these as normal functions instead of extension methods? e.g. `hexChars(s: String)` or `strToByte(s: String)`